src/Aspire.Hosting/ResourceBuilderExtensions.cs

```diff
@@ -1139,6 +1140,289 @@ public static IResourceBuilder<T> WithCommand<T>(
+        string? commandName = null,
+        string? displayDescription = null,
+        string? confirmationMessage = null,
+        string? iconName = null,
+        IconVariant? iconVariant = null,
+        bool isHighlighted = false)
+        where TResource : IResourceWithEndpoints
+    {
+        method ??= HttpMethod.Post;
+
+        endpointSelector ??= PreferredEndpointSelector(builder);
+
+        var endpoint = endpointSelector()
+            ?? throw new DistributedApplicationException($"Could not create HTTP command for resource '{builder.Resource.Name}' as the endpoint selector returned null.");
+
+        commandName ??= $"{endpoint.Resource.Name}-{endpoint.EndpointName}-http-{method.Method.ToLowerInvariant()}";
+
+        var targetRunning = false;
+        builder.ApplicationBuilder.Eventing.Subscribe<BeforeStartEvent>((e, ct) =>
+        {
+            var rns = e.Services.GetRequiredService<ResourceNotificationService>();
+            var watchTask = Task.Run(async () =>
+            {
+                await foreach (var resourceEvent in rns.WatchAsync(ct).WithCancellation(ct))
+                {
+                    if (resourceEvent.Resource == endpoint.Resource)
+                    {
+                        var resourceState = resourceEvent.Snapshot.State?.Text;
+                        targetRunning = resourceState == KnownResourceStates.Running || resourceState == KnownResourceStates.RuntimeUnhealthy;
+                    }
+                }
+            }, ct);
+
+            return Task.CompletedTask;
+        });
+
+        builder.WithCommand(commandName, displayName,
+            async context =>
+            {
+                if (!endpoint.IsAllocated)
+                {
+                    return new ExecuteCommandResult { Success = false, ErrorMessage = "Endpoints are not yet allocated." };
+                }
+
+                var uri = new UriBuilder(endpoint.Url) { Path = path }.Uri;
+                var httpClient = context.ServiceProvider.GetRequiredService<IHttpClientFactory>().CreateClient(httpClientName ?? Options.DefaultName);
+                var request = new HttpRequestMessage(method, uri);
+                if (configureRequest is not null)
+                {
+                    await configureRequest(request).ConfigureAwait(false);
+                }
+                try
+                {
+                    var response = await httpClient.SendAsync(request, context.CancellationToken).ConfigureAwait(false);
+                    if (getCommandResult is not null)
+                    {
+                        return await getCommandResult(response).ConfigureAwait(false);
+                    }
+
+                    response.EnsureSuccessStatusCode();
+                    return CommandResults.Success();
+                }
+                catch (Exception ex)
+                {
+                    return CommandResults.Failure(ex);
```

Comment:
Do we need a test for a failure case?

Reply:
I think we do? See https://github.com/dotnet/aspire/pull/7962/files#diff-39660fc1865df29aaf6dd429fbf374c894ffe5a8330d6660706718438a0758e5R71-R75